cpp/dolfinx/common/array_2d.h

```diff
@@ -0,0 +1,187 @@
+  using size_type = typename std::vector<T, Allocator>::size_type;
+  using reference = typename std::vector<T, Allocator>::reference;
+  using const_reference = typename std::vector<T, Allocator>::const_reference;
+  using pointer = typename std::vector<T, Allocator>::pointer;
+  using iterator = typename std::vector<T, Allocator>::iterator;
+  using const_iterator = typename std::vector<T, Allocator>::const_iterator;
+  /// \endcond
+
+  /// Constructs an empty two dimensional array.
+  array_2d() : _rows(0), _cols(0) { _storage = std::vector<T, Allocator>(); }
+
+  /// Constructs a two dimensional array with size = rows * cols, and initialize
+  /// elements with value value.
+  array_2d(size_type rows, size_type cols, value_type value = T(),
+           const Allocator& alloc = Allocator())
+      : _rows(rows), _cols(cols)
+  {
+    _storage = std::vector<T, Allocator>(_rows * _cols, value, alloc);
+  }
+
+  /// Constructs a two dimensional array with size = rows * cols, and initialize
+  /// elements with value value.
+  template <class Iterator>
+  array_2d(const std::vector<size_t>& shape, Iterator first, Iterator last)
+  {
+    assert(shape.size() == 2);
+    _rows = shape[0];
+    _cols = shape[1];
+    _storage.resize(_rows * _cols);
+    std::copy(first, last, _storage.begin());
+  }
+
+  /// Constructs a two dimensional array with the copy of the contents of other
+  /// two dimensional container.
+  template <class Container,
+            typename
+            = typename std::enable_if<is_2d_container<Container>::value>::type>
+  array_2d(Container& array)
+  {
+    _rows = array.rows();
+    _cols = array.cols();
+    _storage.resize(array.size());
+    std::copy(array.data(), array.data() + array.size(), _storage.begin());
+  }
+
+  /// Constructs a two dimensional array with the contents of the initializer
+  /// list init.
+  array_2d(size_type rows, size_type columns, std::initializer_list<T> list,
+           const Allocator& alloc = Allocator())
+      : _rows(rows), _cols(columns)
+  {
+    if (_rows * _cols != list.size())
+      throw std::runtime_error("Dimension mismatch");
+    _storage = std::vector<T, Allocator>(list, alloc);
+  }
+
+  /// Returns a reference to the element at specified location (i, j).
+  /// No bounds checking is performed.
+  reference operator()(size_type i, size_type j)
+  {
+    return _storage[i * _cols + j];
+  }
+
+  /// Returns a const reference to the element at specified location (i, j).
+  /// No bounds checking is performed.
+  const_reference operator()(size_type i, size_type j) const
+  {
+    return _storage[i * _cols + j];
+  }
+
+  /// Returns a reference to the row at specified location (i).
+  tcb::span<value_type> row(int i) const
+  {
+    size_type offset = i * _cols;
+    return tcb::span<value_type>(pointer(&_storage[0] + offset), _cols);
+  }
+
+  /// Returns a pointer to the first element of the array.
+  value_type* data() noexcept { return _storage.data(); }
```

Comment:
I'd like to put a big warning on the `data` functions in case we introduce offsets to align each row.

Reply:
I think this function is handy for interoperability with eigen and numpy.

For example, iIf we intend to use the buffer protocol to avoid copies, `*ptr=array.data()`
```c++
struct buffer_info {
    void *ptr;
    py::ssize_t itemsize;
    std::string format;
    py::ssize_t ndim;
    std::vector<py::ssize_t> shape;
    std::vector<py::ssize_t> strides;
};
```

Reply:
Agree, but if we have a stride we don't want the function being used carelessly.

Reply:
A warning in the documentation, then?